hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java

```diff
@@ -342,20 +345,25 @@ private Token.Builder customizeToken(
      * @param originalToken original token
      */
     private void updateTokenAttributes(
-            final TokenUpdateTransactionBody op, final Token.Builder builder, final Token originalToken) {
+            final TokenUpdateTransactionBody op,
+            final Token.Builder builder,
+            final Token originalToken,
+            final boolean isHapiCall) {
         if (op.symbol() != null && op.symbol().length() > 0) {
             builder.symbol(op.symbol());
         }
         if (op.name() != null && op.name().length() > 0) {
             builder.name(op.name());
         }
-        if (op.hasMemo()) {
+        if (op.hasMemo() && op.memo().length() > 0) {
             builder.memo(op.memo());
         }
         if (op.hasMetadata()) {
             builder.metadata(op.metadata());
         }
-        if (op.hasTreasury() && !op.treasuryOrThrow().equals(originalToken.treasuryAccountId())) {
+        if (op.hasTreasury()
+                && (isHapiCall || !isZeroAccount(op.treasuryOrThrow()))
+                && !op.treasuryOrThrow().equals(originalToken.treasuryAccountId())) {
```

Comment:
If you can figure out this test (L364-L366) then [_you're a better man than I am, Gunga Din_](https://en.wikipedia.org/wiki/Gunga_Din).

Basic form is: `a && (b || !c) && (!d)` - uh huh.  Oh, and it _requires_ short-circuiting too.

Can't this be fixed up so even I can understand it?  Not just by adding a comment, though that is probably necessary, but by breaking it up and adding meaningful names to subexpressions?